lib/s3routes/routesUtils.ts

```diff
@@ -0,0 +1,1149 @@
+        resources.bucket = undefined;
+    }
+    if (resources.object === '' || resources.object === null) {
+        resources.object = undefined;
+    }
+    return resources;
+}
+
+/**
+ * Get bucket name from the request of a virtually hosted bucket
+ * @param request - HTTP request object
+ * @return - returns bucket name if dns-style query
+ *                              returns undefined if path-style query
+ * @param validHosts - all region endpoints + websiteEndpoints
+ * @throws in case the type of query could not be infered
+ */
+export function getBucketNameFromHost(
+    request: http.IncomingMessage,
+    validHosts: string[],
+) {
+    const headers = request.headers;
+    if (headers === undefined || headers.host === undefined) {
+        throw new Error('bad request: no host in headers');
+    }
+    const reqHost = headers.host;
+    const bracketIndex = reqHost.indexOf(']');
+    const colonIndex = reqHost.lastIndexOf(':');
+
+    const hostLength = colonIndex > bracketIndex ?
+        colonIndex : reqHost.length;
+    // If request is made using IPv6 (indicated by presence of brackets),
+    // surrounding brackets should not be included in host var
+    const host = bracketIndex > -1 ?
+        reqHost.slice(1, hostLength - 1) : reqHost.slice(0, hostLength);
+    // parseIp returns empty object if host is not valid IP
+    // If host is an IP address, it's path-style
+    if (Object.keys(ipCheck.parseIp(host)).length !== 0) {
+        return undefined;
+    }
+
+    let bucketName: string | undefined;
+    for (let i = 0; i < validHosts.length; ++i) {
+        if (host === validHosts[i]) {
+            // It's path-style
+            return undefined;
+        } else if (host.endsWith(`.${validHosts[i]}`)) {
+            const potentialBucketName = host.split(`.${validHosts[i]}`)[0];
+            if (!bucketName) {
+                bucketName = potentialBucketName;
+            } else {
+                // bucketName should be shortest so that takes into account
+                // most specific potential hostname
+                bucketName =
+                    potentialBucketName.length < bucketName.length ?
+                        potentialBucketName : bucketName;
+            }
+        }
+    }
+    if (bucketName) {
+        return bucketName;
+    }
+    throw new Error(
+        `bad request: hostname ${host} is not in valid endpoints`,
+    );
+}
+
+/**
+ * Modify http request object
+ * @param request - http request object
+ * @param validHosts - all region endpoints + websiteEndpoints
+ * @return request object with additional attributes
+ */
+export function normalizeRequest(
+    request: http.IncomingMessage,
+    validHosts: string[],
+) {
+    /* eslint-disable no-param-reassign */
+    const parsedUrl = url.parse(request.url!, true);
+    // @ts-expect-error
+    request.query = parsedUrl.query;
+    // TODO: make the namespace come from a config variable.
```

Comment:
It is a good practice to have TODO attached with a JIRA ticket number to track todos :)